### ISSUE TITLE
Performance opportunity: caching and collection length

### ISSUE THREAD

**Issue by johnkerl** [single-cell-data/TileDB-SOMA#594]
It appears that `__len__` in `soma.Collection` doesn't use the group-member cache, and re-loads from storage every time.

**Comment by bkmartinjr** [comment]
With our current semantics (ie, stateless), caching the group contents leads to bugs - the cache is out of date when the collection is being actively modified.  So I'm not sure this performance optimization is valid as the API currently stands.

Once we have point-in-time open handles to collections, then I fully agree it should "read once"

**Comment by johnkerl** [comment]
@bkmartinjr what we had at one point was:

* Load cache on first access
* Invalidate and reload on any modifying call, e.g. add-member or remove-member

> Once we have point-in-time open handles to collections, then I fully agree it should "read once"

💯

**Comment by bkmartinjr** [comment]
Yep, I know. I removed that caching because it introduced bugs in a multi-writer scenario.  The code assumed single-core code
